File: ios/Sources/NetUtilsPlugin/SSHSession.swift
import Foundation
import Capacitor
import NMSSH

@objc(SSHSession)
public class SSHSession: NSObject, NMSSHChannelDelegate {

  var session: NMSSHSession?
  var channel: NMSSHChannel?

  weak var plugin: NetUtilsPlugin?

  private let bufferQueue = DispatchQueue(label: "com.macchie.SSHSession.buffer")
  private var stdoutBuffer: String = ""
  private var stderrBuffer: String = ""
  private var outputTimer: Timer?

  init(plugin: NetUtilsPlugin) {
    self.plugin = plugin
    super.init()
    startFlushTimer()
  }

  private func startFlushTimer() {
    DispatchQueue.main.async { [weak self] in
      guard let self = self else { return }
      self.outputTimer = Timer.scheduledTimer(withTimeInterval: 0.1, repeats: true) { [weak self] _ in
        self?.flushOutputBuffers()
      }
    }
  }

  func cleanup() {
    outputTimer?.invalidate()
    outputTimer = nil
    if let channel = channel {
      channel.closeShell()
    }
    channel = nil
    session?.disconnect()
    session = nil
  }

  deinit {
    outputTimer?.invalidate()
  }

  public func channel(_ channel: NMSSHChannel!, didReadData message: String!) {
    guard let message = message else { return }
    bufferQueue.async {
      self.stdoutBuffer.append(message)
    }
  }

  public func channel(_ channel: NMSSHChannel!, didReadError error: String!) {
    guard let errorMsg = error else { return }
    bufferQueue.async {
      self.stderrBuffer.append(errorMsg)
    }
  }

  private func flushOutputBuffers() {
    bufferQueue.async { [weak self] in
      guard let self = self else { return }

      if !self.stdoutBuffer.isEmpty {
        let dataToSend = self.stdoutBuffer
        self.stdoutBuffer = ""
        self.plugin?.notifyListeners("ssh:stdout", data: ["data": dataToSend])
      }
      if !self.stderrBuffer.isEmpty {
        let dataToSend = self.stderrBuffer
        self.stderrBuffer = ""
        self.plugin?.notifyListeners("ssh:stderr", data: ["data": dataToSend])
      }
    }
  }

  @objc func connect(_ call: CAPPluginCall) {
    guard let host = call.getString("host"),
          let username = call.getString("username"),
          let password = call.getString("password") else {
      call.reject("Missing parameters. Provide host, username, and password.")
      return
    }

    let port = call.getInt("port") ?? 22

    DispatchQueue.global(qos: .userInitiated).async { [weak self] in
      let session = NMSSHSession.connect(toHost: "\(host):\(port)", withUsername: username)

      if let session = session, session.isConnected {
        session.authenticate(byPassword: password)
        if session.isAuthorized {
          self?.session = session
          call.resolve(["success": true])
        } else {
          session.disconnect()
          call.reject("Authentication failed")
        }
      } else {
        call.reject("Connection failed")
      }
    }
  }

  @objc func startShell(_ call: CAPPluginCall) {
    guard let session = session, session.isConnected else {
      call.reject("No active session. Call connect() first.")
      return
    }

    channel = session.channel
    channel?.delegate = self
    channel?.requestPty = true

    DispatchQueue.global(qos: .background).async { [weak self] in
      do {
        try self?.channel?.startShell()
        call.resolve(["success": true])
      } catch let error {
        call.reject("Failed to start shell: \(error.localizedDescription)")
      }
    }
  }

  @objc func write(_ call: CAPPluginCall) {
    guard let command = call.getString("command") else {
      call.reject("No command provided")
      return
    }

    guard let channel = channel else {
      call.reject("No active shell. Call startShell() first.")
      return
    }

    DispatchQueue.global(qos: .userInitiated).async {
      do {
        try channel.write(command + "\n")
        call.resolve(["success": true])
      } catch let error {
        call.reject("Failed to write command: \(error.localizedDescription)")
      }
    }
  }

  @objc func disconnect(_ call: CAPPluginCall) {
    cleanup()
    call.resolve(["success": true])
  }
}

File: android/src/main/java/com/macchie/plugins/netutils/SSHSession.java
package com.macchie.plugins.netutils;

import com.getcapacitor.JSObject;
import com.getcapacitor.PluginCall;
import com.jcraft.jsch.*;

import java.io.InputStream;
import java.io.OutputStream;
import java.util.concurrent.Executors;

public class SSHSession {

  private Session jschSession;
  private Channel shellChannel;
  private OutputStream shellInput;
  private final NetUtilsPlugin plugin;

  public SSHSession(NetUtilsPlugin plugin) {
    this.plugin = plugin;
  }

  public void connect(PluginCall call) {
    String host = call.getString("host");
    String username = call.getString("username");
    String password = call.getString("password");
    int port = call.getInt("port", 22);

    if (host == null || username == null || password == null) {
      call.reject("Missing parameters. Provide host, username, and password.");
      return;
    }

    Executors.newSingleThreadExecutor().execute(() -> {
      try {
        JSch jsch = new JSch();
        jschSession = jsch.getSession(username, host, port);
        jschSession.setPassword(password);
        jschSession.setConfig("StrictHostKeyChecking", "no");
        jschSession.connect(10000);

        if (jschSession.isConnected()) {
          JSObject result = new JSObject();
          result.put("success", true);
          call.resolve(result);
        } else {
          call.reject("Connection failed");
        }
      } catch (Exception e) {
        call.reject("Connection failed: " + e.getMessage());
      }
    });
  }

  public void startShell(PluginCall call) {
    if (jschSession == null || !jschSession.isConnected()) {
      call.reject("No active session. Call connect() first.");
      return;
    }

    Executors.newSingleThreadExecutor().execute(() -> {
      try {
        shellChannel = jschSession.openChannel("shell");
        InputStream in = shellChannel.getInputStream();
        shellInput = shellChannel.getOutputStream();
        shellChannel.connect();

        JSObject result = new JSObject();
        result.put("success", true);
        call.resolve(result);

        // Read loop in background
        Executors.newSingleThreadExecutor().execute(() -> {
          byte[] buffer = new byte[1024];
          int len;
          try {
            while ((len = in.read(buffer)) != -1) {
              String data = new String(buffer, 0, len);
              JSObject event = new JSObject();
              event.put("data", data);
              plugin.emit("ssh:stdout", event);
            }
          } catch (Exception ignored) {
            // Connection closed
          }
        });

      } catch (Exception e) {
        call.reject("Failed to start shell: " + e.getMessage());
      }
    });
  }

  public void write(PluginCall call) {
    String command = call.getString("command");
    if (command == null) {
      call.reject("No command provided");
      return;
    }

    if (shellInput == null) {
      call.reject("No active shell. Call startShell() first.");
      return;
    }

    Executors.newSingleThreadExecutor().execute(() -> {
      try {
        shellInput.write((command + "\n").getBytes());
        shellInput.flush();
        JSObject result = new JSObject();
        result.put("success", true);
        call.resolve(result);
      } catch (Exception e) {
        call.reject("Failed to write command: " + e.getMessage());
      }
    });
  }

  public void disconnect(PluginCall call) {
    try {
      if (shellChannel != null) {
        shellChannel.disconnect();
        shellChannel = null;
        shellInput = null;
      }
      if (jschSession != null && jschSession.isConnected()) {
        jschSession.disconnect();
        jschSession = null;
      }
      JSObject result = new JSObject();
      result.put("success", true);
      call.resolve(result);
    } catch (Exception e) {
      call.reject("Failed to disconnect: " + e.getMessage());
    }
  }
}
